src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -720,8 +748,32 @@ where
 
                 let mut batch_output = batch_output.activate();
                 let mut summary_output = summary_output.activate();
+                let mut health_output = health_output.activate();
 
                 for (message_batch, source_upper) in buffer.drain(..) {
+                    if let Some(batch) = &*message_batch.borrow() {
+                        let has_errors = batch.non_definite_errors.first();
+                        let has_messages = batch.messages.values().any(|vs| !vs.is_empty());
+
+                        let maybe_health = match (has_errors, has_messages) {
+                            (Some(error), _) => {
```

Comment:
musing: should we maybe not allow a source to emit both updates and errors? It's hard to reason about whether we're in a consistent state when we allow those messages to go downstream and be persisted. But, idk...